### PR TITLE
`within group` parser support

### PR DESCRIPTION
**Pull request by kovrus**

### CODE REVIEW

File: sql-parser/src/main/java/io/crate/sql/tree/WithinGroup.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class WithinGroup extends Node {
+
+    private final SortItem orderBy;
+
+    public WithinGroup(SortItem orderBy) {
```
Comment (matriv) [inline sql-parser/src/main/java/io/crate/sql/tree/WithinGroup.java]: You might as well add the @Nonnnull to the orderBy arg.

File: sql-parser/src/main/java/io/crate/sql/ExpressionFormatter.java [hunk 2]
```diff
@@ -236,29 +237,38 @@ public String visitInputReference(InputReference node, Void context)
...
-            if (node.isDistinct()) {
+            if (node.distinct()) {
                 arguments = "DISTINCT " + arguments;
             }
 
-            builder.append(node.getName())
-                    .append('(').append(arguments).append(')');
+            builder.append(node.name()).append('(').append(arguments).append(")");
+
+            Optional<Window> window = node.window();
```
Comment (matriv) [inline sql-parser/src/main/java/io/crate/sql/ExpressionFormatter.java]: I would skip the local var

File: sql-parser/src/main/java/io/crate/sql/ExpressionFormatter.java [hunk 3]
```diff
@@ -236,29 +237,38 @@ public String visitInputReference(InputReference node, Void context)
...
+            if (window.isPresent()) {
+                builder.append(" OVER ").append(visitWindow(window.get(), null));
+            }
 
-            if (node.getWindow().isPresent()) {
-                builder.append(" OVER ").append(visitWindow(node.getWindow().get(), null));
+            Optional<WithinGroup> withinGroup = node.withinGroup();
```
Comment (matriv) [inline sql-parser/src/main/java/io/crate/sql/ExpressionFormatter.java]: same here

File: sql-parser/src/main/java/io/crate/sql/ExpressionFormatter.java [hunk 4]
```diff
@@ -236,29 +236,36 @@ public String visitInputReference(InputReference node, Void context)
...
             StringBuilder builder = new StringBuilder();
 
-            String arguments = joinExpressions(node.getArguments());
-            if (node.getArguments().isEmpty() && "count".equalsIgnoreCase(node.getName().getSuffix())) {
+            String arguments = joinExpressions(node.arguments());
+            if (node.arguments().isEmpty() && "count".equalsIgnoreCase(node.name().getSuffix())) {
                 arguments = "*";
             }
-            if (node.isDistinct()) {
+            if (node.distinct()) {
```
Comment (mfussenegger) [inline sql-parser/src/main/java/io/crate/sql/ExpressionFormatter.java]: why did you remove the `is` prefix?

File: sql-parser/src/main/java/io/crate/sql/tree/WithinGroup.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        if ((obj == null) || (getClass() != obj.getClass())) {
+            return false;
+        }
+        WithinGroup o = (WithinGroup) obj;
+        return Objects.equal(orderBy, o.orderBy);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(orderBy);
```
Comment (mfussenegger) [inline sql-parser/src/main/java/io/crate/sql/tree/WithinGroup.java]: I think orderBy.hashCode could be used directly instead of wrapping it into a `Objects.hashCode` call. Same for equals.

**Comment by matriv** [comment]
Igtm
